### PR TITLE
Bigquery output Plugin

### PR DESCRIPTION
**Pull request by gkatzioura**
### Required for all PRs:

The pull request for the Issue #8630

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

File: plugins/outputs/bigquery/bigquery.go
Comment (Jagularr) [inline plugins/outputs/bigquery/bigquery.go:61]: Feels like this method could be moved to the test file

**Review by Jagularr — COMMENTED** [review]
Looks solid to me

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:10]: You might want to put this one in a separate "group" below as it is an "internal" package.

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:73]: Can you please move this to the testing! You can guard the actual function by a check for `b.client` not being `nil` if you have to.

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:48]: For testing you could do
```suggestion
	// To allow testing we do not override the client. The client will never be initialized in
	// normal environment. 
	if b.client == nil {
		b.setUpClient()
	}
```
and init `b.Client` with what you have in `setUpTestClient()`. This way you can move all

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:72]: As you return in the if you can save some indention:
```suggestion
	}
	b.client = c

	return nil
```

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:78]: Maybe you want a user-definable timeout and use it here!? Do not forget to `defer cancel()` in this case.

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:103]: This means you drop all remaining metrics. Skipping over this erroneous one but process the others might be better
```suggestion
			b.Log.Errorf("inserting metric %q failed: %v", <metric id/name>, err)
			continue
```

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:102]: Does it make sense to do this in parallel using a go-function?

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:190]: Again, you might want a user-definable timeout...

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:215]: Usually those are at the top.

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:206]: You are missing a `init()` function to register this output plugin.

**Review by srebhan — CHANGES_REQUESTED** [review]
Hey this looks quite good already. My main concerns are

1. You are missing an `init()` function to register the plugin.
2. You might want a user-definable `timeout` option to avoid blocking forever in operations.
3. You should move the testing machinery out of the main plugin file.
4. Please add `cloud.google.com/go/bigquery` to `docs/LICENSE_OF_DEPENDENCIES.md`.

See also my comments in the code.

File: plugins/outputs/bigquery/bigquery.go
Comment (gkatzioura) [reply plugins/outputs/bigquery/bigquery.go]: I updated it, to achieve it through a Wait Group.

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:29]: Could you please avoid abbreviations here as some first-time-users might have difficulties to understand what it is about... :-)

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:83]: Could you please also move this to `Connect()` to have a nice consistent structure. It will add to the beauty of this code.

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:35]: Is there a reason you enforce a timeout of one second? If it is about default-value handling, the easy solution is to set the default in `init()` when initializing the structure. This way the value is only overwritten when set by the user.

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:117]: You might use an anonymous function here to put the `defer wg.Done()` here rather than the function body for everybody to see it's done correctly. But that is a matter of taste and I leave it to you.

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [reply plugins/outputs/bigquery/bigquery.go:208]: Please mark as resolved.

File: plugins/outputs/bigquery/bigquery_test.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery_test.go:33]: I don't get the intention behind this test... If you only want to test if you can setup a local test-server, I think you don't need to do it.

File: plugins/outputs/bigquery/bigquery_test.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery_test.go:59]: Could you add a test where you actually compare what is sent to the server with what you expect? Do you have an example of the on-wire format?

**Review by srebhan — COMMENTED** [review]
Very nice code! Thank you! Only few comments in the code. One nice-to-have thing is a test where you actually compare what is sent to the server with your expectations. But I'm not sure if this is doable!?

File: plugins/outputs/bigquery/bigquery.go
Comment (gkatzioura) [reply plugins/outputs/bigquery/bigquery.go]: Moved to anonymous function

**Comment by srebhan** [comment]
Hey @gkatzioura you are writing the code faster than I can review! ;-) Good job! :-)

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:115]: Make it `defer wg.Done()` and move it to the top of the anonymous function so we can be sure it will be done in any case.

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [reply plugins/outputs/bigquery/bigquery.go:84]: Please mark as resolved.

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:116]: This should also work
```suggestion
		go func() {
			defer wg.Done()
			b.insertToTable(k, v)
		}()
```

**Comment by srebhan** [comment]
@gkatzioura can you please comment on why you enforce a minimum timeout of 1s!?

**Comment by gkatzioura** [comment]
> @gkatzioura can you please comment on why you enforce a minimum timeout of 1s!?

Hi @srebhan will do, bit busy now due to work.

**Comment by srebhan** [comment]
@gkatzioura ok don't worry, just liked the live coding/review in the morning... ;-)

File: plugins/outputs/bigquery/bigquery.go
Comment (gkatzioura) [reply plugins/outputs/bigquery/bigquery.go:35]: Hi @srebhan no specific reason. Was just checking the zookeeper connector that did the same. Did the default as you proposed.

File: plugins/outputs/bigquery/bigquery_test.go
Comment (gkatzioura) [reply plugins/outputs/bigquery/bigquery_test.go:36]: I run the tests through TestMain. I use TestMain in order to spin up the server that produces a fake BigQuery Server. Then through T.Run() the tests `TestWrite` and  `TestConnect` will run using the fake BigQuery Server.

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [reply plugins/outputs/bigquery/bigquery.go]: Please mark as resolved.

File: plugins/outputs/bigquery/bigquery_test.go
Comment (srebhan) [reply plugins/outputs/bigquery/bigquery_test.go:36]: This is an unusual pattern for telegraf. :-) Is there any reason not to spin up a server in every test (that's what everybody else in telegraf does)?

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:24]: Can you please do:
```suggestion
var defaultTimeout = internal.Duration{Duration: 5 * time.Second}
```
as there is a big PR fixing those superfluous expressions (#8835) and Pawel will not be happy if we add more. ;-)

File: plugins/outputs/bigquery/bigquery_test.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery_test.go:26]: Please remove `time.Duration`.

**Comment by srebhan** [comment]
Almost. Please remove the `time.Duration` as the timeout is already a duration. Furthermore, you can remove`cloud.google.com/go/bigquery` from the list-of-dependencies document as `cloud.google.com/go/*` is covered already.

Last thing is the test-structure. Can you please start the server in each test individually to match "the usual pattern"!?

File: plugins/outputs/bigquery/bigquery_test.go
Comment (gkatzioura) [reply plugins/outputs/bigquery/bigquery_test.go]: Changed to spinning up per test.

File: plugins/outputs/bigquery/README.md
Comment (srebhan) [inline plugins/outputs/bigquery/README.md:32]:
```suggestion
Avoid hyphens on BigQuery tables, underlying SDK cannot handle streaming inserts to Table with hyphens.
```?

File: plugins/outputs/bigquery/README.md
Comment (srebhan) [inline plugins/outputs/bigquery/README.md:40]:
```suggestion
All field naming restrictions that apply to BigQuery should apply to the measurements to be imported.
```

File: plugins/outputs/bigquery/README.md
Comment (srebhan) [inline plugins/outputs/bigquery/README.md:44]:
```suggestion
Pay attention to the column `timestamp` since it is reserved upfront and cannot change. 
```?

File: plugins/outputs/bigquery/README.md
Comment (srebhan) [inline plugins/outputs/bigquery/README.md:42]:
```suggestion
Tables on BigQuery should be created beforehand and they are not created during persistence
```

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:22]: I think this isn't used.
```suggestion
const timeStampFieldName = "timestamp"
```

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:29]:
```suggestion
  credentials_file = "/path/to/service/account/key.json"

  ## Google Cloud Platform Project
```

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:43]:
```suggestion
```

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:46]:
```suggestion
	Log telegraf.Logger `toml: "-"`
```

**Review by srebhan — COMMENTED** [review]
I went through everything a (almost) last time and found some more typos and formatting things. Only cosmetics. After fixing those I think we are good to go.

**Comment by gkatzioura** [comment]
Hi @srebhan I added a map in order to do a lookup for cases of metrics with hyphens. Due to the SDK and the api having issues inserting to tables with hyphens this connector would have been limited to only non-hyphen metrics. 
Instead by having this map the user can define an association between a metric with hyphens and the corresponding table.

**Comment by srebhan** [comment]
I'm a bit ambivalent here. On the one hand-side I think letting the user specify a replacement (e.g. underscore, space or nothing, i.e. empty) would be good enough in my view, on the other hand this is not much code. What makes me tend to the "user specifies a replacement" (with a sensible default) solution is the fact that this way you can _ensure_ that you can process the metric.
In the table case, the user might simply forget, mistype or do some other error that leaves you with a hyphen-metric without a mapping in the table. What do you do in such cases?

Any thoughts?

**Comment by gkatzioura** [comment]
Hi @srebhan. 
My concern was users dropping the plugin due to the naming limitation, since many measurement will have hyphens.If the users specify a map (for the hyphens cases, there is no need to do it for everything, the connector should also work out of the box) they are covered on this issue but as you mentioned they are not covered in case they forget that they have a hyphen metric.
It seems ok in case of a hyphen metric where the user has not specified a table mapping, to fallback on removing the hyphens (and print a warning). 
Thus co-existence of table mapping exists and also a fallback.
However the user still needs to create the BigQuery tables upfront. So somehow he still needs to be aware on how to name the tables and there also might be a rename collision.

**Comment by srebhan** [comment]
@gkatzioura I can follow your argumentation, but a mixture of concepts (replacement of hyphen vs. replacement of metric name) is always the worst thing to do... ;-)

> If the users specify a map (for the hyphens cases, there is no need to do it for everything, the connector should also work out of the box) they are covered on this issue but as you mentioned they are not covered in case they forget that they have a hyphen metric.

Right. That's my concern. In large installations with many plugins and maybe multiple users modifying the config you will hit the  "forgotten mapping" case with certainty! That's why I would prefer to specify a replacement character (e.g. underscore) for the hyphens. If the user wants to really rename the metrics he/she can still use the [rename processor](https://github.com/influxdata/telegraf/tree/master/plugins/processors/rename) for doing so. This would keep the config options for your plugins short (it's just so beautifully brief!) and makes sure we cover _all_ cases by default.

**Comment by gkatzioura** [comment]
Hi @srebhan I think the [rename processor](https://github.com/influxdata/telegraf/tree/master/plugins/processors/rename) is amazing and there no need for what I did with the map (was not fully aware :/ on it). Thus I shall remove the map mechanism.

On your comment you proposed to swap hyphens with underscores. Are you ok for me to proceed?

**Comment by srebhan** [comment]
@gkatzioura yeah that's my proposal. However, I would like to see a config option with the replacement, having underscore as default and removing hyphens if empty. Do you think this makes sense?

For logging on replace, please track which metric you already warned on to avoid spamming the log-file. You can do this easily by using a `map[string]bool` that you set if you warned and before warning check `if ! alreadyWarned[metric]`...

**Comment by gkatzioura** [comment]
Hi @srebhan will proceed as suggested.

**Comment by gkatzioura** [comment]
Hi @srebhan did proceed on the changes and updated the Readme.
* Added a default hyphen replacement
* Added configuration on the character replacing
* Print warning only once

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:36]: This misses the `replace_hyphen_to` option. Please also sync with `README.md`.

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:228]: Please invert the logic to save indention levels:
```suggestion
	if ! strings.Contains(metricName, "-") {
		return metricName
	}

	dhm := strings.ReplaceAll(metricName, "-", b.ReplaceHyphenTo)
	if warned, _ := b.warnedOnHyphens[metricName]; !warned {
		b.Log.Warn("Metric %q contains hyphens please consider using the rename processor plugin, falling back to %q", dhm)
		b.warnedOnHyphens[metricName] = true
	}

	return dhm
```

File: plugins/outputs/bigquery/bigquery.go
Comment (srebhan) [inline plugins/outputs/bigquery/bigquery.go:244]: Please do this in `Init()`.

**Review by srebhan — COMMENTED** [review]
Very good, please check the minor comments and update `sampleConfig` and `README.md` to the new option(s).

File: plugins/outputs/bigquery/bigquery.go
Comment (gkatzioura) [reply plugins/outputs/bigquery/bigquery.go]: I moved this above. Did you mean doing the map initialization on the Connect()?

Comment (srebhan) [reply plugins/outputs/bigquery/bigquery.go]: I mean moving `b.warnOnHyphens = make(map[string]bool)` to `Connect()` instead of doing it here. Sorry for the confusion, `Init()` is the `Connect()` in input plugins... :-)

**Review by srebhan — APPROVED** [review]
This code looks beautiful IMO. Thank you very much for your effort!

**Comment by srebhan** [comment]
Oh just one more request @gkatzioura... Please add this plugin to the [top-level README file's output list](https://github.com/influxdata/telegraf/blob/master/README.md#output-plugins).

**Review by srebhan — APPROVED** [review]
Perfect.

**Comment by srebhan** [comment]
@gkatzioura please resolve the merge conflicts!

**Review by srebhan — APPROVED** [review]
It's off again, please take a look from time to time so this one stays merge-able. I'll try to find out when this might be merged.

**Comment by ssoroka** [comment]
I'm good to merge this after you've had a look at my last two comments.

File: plugins/outputs/bigquery/bigquery.go
Comment (gkatzioura) [reply plugins/outputs/bigquery/bigquery.go:217]: Yes this is a no retry case. In case of a failure due to the availability of BigQuery the assumption is that the connector is misconfigured (BigQuery Table does not exist, wrong fields). Let me know what would work better.

Comment (ssoroka) [reply plugins/outputs/bigquery/bigquery.go]: I'm fine with it if you are. I do wonder if maybe the Goroutines aren't really necessary. The retry logic would be easier without them, as on 5xx status code you'd return an error, and on 4xx you'd log it and return nil. If they're all going over the same client connection, I think they're being serialized anyway.

**Comment by ssoroka** [comment]
I'm going to merge and you can adjust the goroutines if you want in a new PR. [See my last reply](https://github.com/influxdata/telegraf/pull/8634#discussion_r593324308)

File: plugins/outputs/bigquery/bigquery.go
Comment (gkatzioura) [reply plugins/outputs/bigquery/bigquery.go]: Yes, initially the pr was not using Goroutines, but I was encouraged to parallelize the inserts. I will raise a PR removing them, thus more control and will return on error.

**Pull request merged**
